File: CustomApplications/CSharp/GraphicsHowTo/ScreenOverlays/OverlaysMemoryCodeSnippet.cs
#region UsingDirectives
using System.Drawing;
using System.Collections.Generic;
using AGI.STKGraphics;
using AGI.STKObjects;
using System;
using System.Drawing.Imaging;
using System.Runtime.InteropServices;
#endregion

namespace GraphicsHowTo.ScreenOverlays
{
    class OverlaysMemoryCodeSnippet : CodeSnippet
    {
        public OverlaysMemoryCodeSnippet()
            : base(@"ScreenOverlays\OverlaysMemoryCodeSnippet.cs")
        {
        }

        public override void Execute(IAgStkGraphicsScene scene, AgStkObjectRoot root)
        {
            string imageFile = new AGI.DataPath(AGI.DataPathRoot.Relative, "OverlayFromMemory.bmp").FullPath;
            Execute(scene, root, imageFile);
        }

        public void Execute([AGI.CodeSnippets.CodeSnippet.Parameter("Scene", "Current Scene")] IAgStkGraphicsScene scene, [AGI.CodeSnippets.CodeSnippet.Parameter("Root", "STK Object Model root")] AgStkObjectRoot root, [AGI.CodeSnippets.CodeSnippet.Parameter("imageFile", "The image file")] string imageFile)
        {
#region CodeSnippet
            IAgStkGraphicsSceneManager manager = ((IAgScenario)root.CurrentScenario).SceneManager;
            IAgStkGraphicsScreenOverlayCollectionBase overlayManager = (IAgStkGraphicsScreenOverlayCollectionBase)manager.ScreenOverlays.Overlays;

            int size = /*$size$The size of the screen overlay$*/256;
            ProcedurallyGeneratedTexture tex = new ProcedurallyGeneratedTexture(size);

            // Create a Bitmap and BitmapData and Lock all pixels to be written
            Bitmap bitmap = new Bitmap(size, size);
            BitmapData bitmapData = bitmap.LockBits(
                                       new Rectangle(0, 0, bitmap.Width, bitmap.Height),
                                       ImageLockMode.WriteOnly, bitmap.PixelFormat);

            // Copy the data from the byte array into BitmapData.Scan0
            byte[] data = tex.Next();
            Marshal.Copy(data, 0, bitmapData.Scan0, data.Length);

            // Unlock the pixels and save temporarily
            bitmap.UnlockBits(bitmapData);
            bitmap.Save(imageFile);

            IAgStkGraphicsRaster img = manager.Initializers.Raster.InitializeWithStringUriXYWidthAndHeight(
                imageFile, 0, 0, size, size);

            IAgStkGraphicsTextureScreenOverlay overlay = manager.Initializers.TextureScreenOverlay.InitializeWithXYWidthHeight(0, 0, size, size);
            overlay.Texture = manager.Textures.FromRaster((IAgStkGraphicsRaster)img);
            ((IAgStkGraphicsOverlay)overlay).Origin = AgEStkGraphicsScreenOverlayOrigin.eStkGraphicsScreenOverlayOriginCenter;

            overlayManager.Add((IAgStkGraphicsScreenOverlay)overlay);
            scene.Render();
#endregion
            m_Overlay = (IAgStkGraphicsScreenOverlay)overlay;
            System.IO.File.Delete(imageFile);
        }

        public override void View(IAgStkGraphicsScene scene, AgStkObjectRoot root)
        {
            // Use current view.
        }

        public override void Remove(IAgStkGraphicsScene scene, AgStkObjectRoot root)
        {
            IAgStkGraphicsSceneManager manager = ((IAgScenario)root.CurrentScenario).SceneManager;
            IAgStkGraphicsScreenOverlayCollectionBase overlayManager = (IAgStkGraphicsScreenOverlayCollectionBase)manager.ScreenOverlays.Overlays;

            overlayManager.Remove(m_Overlay);
            scene.Render();

            m_Overlay = null;
        }

        private IAgStkGraphicsScreenOverlay m_Overlay;
    };
}

File: CustomApplications/VB.Net/GraphicsHowTo/ScreenOverlays/OverlaysMemoryCodeSnippet.vb
#Region "UsingDirectives"
Imports System.Drawing
Imports System.Collections.Generic
Imports AGI.STKGraphics
Imports AGI.STKObjects
Imports System.Drawing.Imaging
Imports System.Runtime.InteropServices
#End Region

Namespace ScreenOverlays
	Class OverlaysMemoryCodeSnippet
		Inherits CodeSnippet
		Public Sub New()
            MyBase.New("ScreenOverlays\OverlaysMemoryCodeSnippet.vb")
		End Sub

        Public Overrides Sub Execute(<AGI.CodeSnippets.CodeSnippet.Parameter("Scene", "Current Scene")> ByVal scene As IAgStkGraphicsScene, <AGI.CodeSnippets.CodeSnippet.Parameter("Root", "STK Object Model root")> ByVal root As AgStkObjectRoot)
            '#Region "CodeSnippet"
            Dim manager As IAgStkGraphicsSceneManager = DirectCast(root.CurrentScenario, IAgScenario).SceneManager
            Dim overlayManager As IAgStkGraphicsScreenOverlayCollectionBase = DirectCast(manager.ScreenOverlays.Overlays, IAgStkGraphicsScreenOverlayCollectionBase)

            Dim size As Integer = attachID("$size$The size of the screen overlay$", 256)
            Dim tex As New ProcedurallyGeneratedTexture(size)

            ' Create a Bitmap and BitmapData and Lock all pixels to be written
            Dim bitmap As New Bitmap(size, size)
            Dim bitmapData As BitmapData = bitmap.LockBits(New Rectangle(0, 0, bitmap.Width, bitmap.Height), ImageLockMode.WriteOnly, bitmap.PixelFormat)

            ' Copy the data from the byte array into BitmapData.Scan0
            Dim data As Byte() = tex.Next()
            data = tex.Next()
            Marshal.Copy(data, 0, bitmapData.Scan0, data.Length)

            ' Unlock the pixels and save temporarily
            bitmap.UnlockBits(bitmapData)
            bitmap.Save(imagePath)

            Dim img As IAgStkGraphicsRaster = manager.Initializers.Raster.InitializeWithStringUriXYWidthAndHeight(imagePath, 0, 0, size, size)

            Dim overlay As IAgStkGraphicsTextureScreenOverlay = manager.Initializers.TextureScreenOverlay.InitializeWithXYWidthHeight(0, 0, size, size)
            overlay.Texture = manager.Textures.FromRaster(DirectCast(img, IAgStkGraphicsRaster))
            DirectCast(overlay, IAgStkGraphicsOverlay).Origin = AgEStkGraphicsScreenOverlayOrigin.eStkGraphicsScreenOverlayOriginCenter

            overlayManager.Add(DirectCast(overlay, IAgStkGraphicsScreenOverlay))
            scene.Render()
            '#End Region
            m_Overlay = DirectCast(overlay, IAgStkGraphicsScreenOverlay)
            System.IO.File.Delete(imagePath)
        End Sub

		Public Overrides Sub View(scene As IAgStkGraphicsScene, root As AgStkObjectRoot)
			' Use current view.
		End Sub

		Public Overrides Sub Remove(scene As IAgStkGraphicsScene, root As AgStkObjectRoot)
			Dim manager As IAgStkGraphicsSceneManager = DirectCast(root.CurrentScenario, IAgScenario).SceneManager
			Dim overlayManager As IAgStkGraphicsScreenOverlayCollectionBase = DirectCast(manager.ScreenOverlays.Overlays, IAgStkGraphicsScreenOverlayCollectionBase)

			overlayManager.Remove(m_Overlay)
			scene.Render()

			m_Overlay = Nothing
		End Sub

		Private m_Overlay As IAgStkGraphicsScreenOverlay
		Private imagePath As String = New AGI.DataPath(AGI.DataPathRoot.Relative, "OverlayFromMemory.bmp").FullPath
	End Class
End Namespace
